### ISSUE TITLE
concept support

### ISSUE THREAD

**Issue by marehr** [include-what-you-use/include-what-you-use#917]
Hi all,

this is my first time experimenting with iwyu and I don't know if I did something wrong, but the following example does not work for me:

```c++
#include <concepts>

static_assert(std::integral<int>);

template <std::integral t>
struct foo
{
    t bar;
};

int main()
{
    std::integral auto bar = 0;
    foo<int> f{bar};
    return f.bar;
}
```

output is
```bash
> include-what-you-use --version
include-what-you-use 0.15 based on clang version 11.1.0

> include-what-you-use --std=c++20 test.cpp

test.cpp should add these lines:

test.cpp should remove these lines:
- #include <concepts>  // lines 1-1

The full include-list for test.cpp:
---
```

**Comment by kimgr** [comment]
The IWYU AST traversal does not look out for concept constructs, so that's probably why they aren't picked up.

I haven't read up on concepts yet, so the syntax looks a little weird to me. I'm guessing `std::integral` is provided by `<concepts>`, but what do the different mentions of `std::integral` mean? The one in `main` looks particularly odd :-)

**Comment by marehr** [comment]
Okay, let me try to explain it :)

> I haven't read up on concepts yet, so the syntax looks a little weird to me. I'm guessing `std::integral` is provided by `<concepts>`

Yes, it is defined as

```c++
template < class T >
concept integral = std::is_integral_v<T>;
```

https://en.cppreference.com/w/cpp/concepts/integral

So it is basically the same as `std::is_integral_v`, but declared as concept.

```c++
static_assert(std::integral<int>);
// is the same as
static_assert(std::is_integral_v<int>);
// but gives better diagnostics
```

See https://godbolt.org/z/53cvWrrjx

```c++
template <std::integral t> // short-syntax
struct foo
{
    t bar;
};
```

is the same as 

```c++
template <typename t>
  requires std::integral<t> // this is the more verbose syntax
struct foo
{
    t bar;
};
```

The second one show the nature of that construct better, It's basically a template class with the extra constraint on the template parameter `t` to be an integral. You can imagine that this can be arbitrarily complicated. For example, you could ensure that a given template parameter behaves like a std container, e.g. has an insert, delete and so on member.

The concept part will provide you with a diagnostic if `t` doesn't fulfil that.

```c++
using foo_int = foo<int>; // will work
using foo_int = foo<float>; // will give an error that float isn't an integral
```

https://godbolt.org/z/zvqnEj9es

And the last syntax 

```c++
int main()
{
    std::integral auto bar = 0;
    foo<int> f{bar};
    return f.bar;
}
````

is kind of the same as just writing

```c++
int main()
{
    /*std::integral*/ auto bar = 0;
    // basically the same as
    static_assert(std::integral<decltype(bar)>);
    foo<int> f{bar};
    return f.bar;
}
````

but ensures that the type of your auto-initialised variable will be of a type that models std::integral.

**Comment by kimgr** [comment]
Thanks for the background!

If you run IWYU with `-Xiwyu -v5`, you'll see the AST being dumped. There seems to be a few constructs we don't recognize there:

* `ConceptSpecializationExpr` in the static assert seems to point back to `std::integral`
* Likewise `ConceptSpecializationExpr` in the template parameter
* The `VarDecl` for `bar` has an `AutoType` with a type constraint that also points back to `std::integral`

So if these AST nodes were properly handled, we should be able to point back to `<concepts>` correctly.

**Comment by marehr** [comment]
This sounds awesome, but I'm not too sure how hard it is to implement that handling. What needs to be done to add these nodes?

**Comment by kimgr** [comment]
@marehr The `RecursiveASTVisitor` implementation in `iwyu.cc` has lots of code to visit AST nodes and figure out what context they're in to be able to report uses, so something needs to be added there. But then there's templates, etc...  :-)

**Comment by captaincrutches** [comment]
It's also worth noting here that a concept can have other types as template arguments, like `std::derived_from` or `std::constructible_from`, and in those cases you often (always?) need the complete type you're talking about - so for IWYU purposes, if you say something like
```cpp
template <std::derived_from<Foo> T>
void doSomething(T& t) { ... }
```
then you need the complete type `Foo` included in that TU; a forward declaration won't cut it.

For example:  a header `mytype.h` which contains only
```cpp
struct MyType {};
```
included by a source file that uses it in a concept such as:
```cpp
#include <concepts>

#include "mytype.h"

static_assert(!std::constructible_from<MyType, int>);

int main() {}
```
IWYU suggests that `<concepts>` be removed (it can't because that's where `std::constructible_from` comes from), and also that `mytype.h` be replaced with a forward declaration of `MyType` (it can't because `std::constructible_from` needs a complete type).

In a C++20 project that uses concepts (and ranges, many of which are also concepts) extensively, this can potentially result in a lot of false warnings.

**Comment by kimgr** [comment]
@marehr The original case in https://github.com/include-what-you-use/include-what-you-use/issues/917#issue-892015959 appears to be fixed on master.

@captaincrutches The example you provided still suggests that `MyType` is made a forward decl, but I believe that's different enough that it's a separate issue (and probably a duplicate of #1414). Let me know if you agree, then we can close this issue.

**Comment by captaincrutches** [comment]
Agreed, #1414 does seem to be the same issue as my example.  I'm fine with rolling into that one.  Thanks!
